---
title: "Bsub PCA"
author: "Yen-Kai Chen"
date: "20/04/2020"
output: word_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# creating the required file
```{r}
rm(list = ls())
setwd("C:/Users/strki/Documents/Git_R/GLDS-138 metabolomics/GLDS-138-metabolomics/Compre Bsub PCA")

# file already downloaded
Bdata <- read.csv("GLDS-138_metabolomics_mx 367428_NASA_bacteria cells_09-2017_submit.csv", stringsAsFactors = FALSE)

# filtering data (Samples in rows (unpaired))
Bdata <- t(Bdata)
Bdata <- Bdata[-c(2:7, 114:118),]
Bdata <- Bdata[,-c(1:2,6:8)]
Bdata[1,1:3] <- Bdata[2,1:3]
Bdata <- Bdata[-c(2),]

rownames(Bdata) <- NULL

Bdata[1,1] <- "Sample"   # renaming to suit MetaboAnalyst
Bdata[1,2] <- "Label"

# Giving column and row names
columnName <- Bdata[1,]
colnames(Bdata) <- columnName
Bdata <- Bdata[-1,]
Bdata <- as.data.frame(Bdata)

# as.numeric
Bdata[,4:ncol(Bdata)] <- lapply(Bdata[,4:ncol(Bdata)], function(x) as.numeric(as.character(x)))

# Row-wise correction
BdataRWC <- as.data.frame(Bdata)
rowi <- BdataRWC[,4:ncol(BdataRWC)]

BdataRWC[,4:ncol(BdataRWC)] <- rowi / rowSums(rowi) # by the peak of all sums

sum(BdataRWC[105,4:ncol(BdataRWC)]) # checking the row wise correction worked well

# Baseline correction
tail(BdataRWC[,1:5], 1)

blc <- tail(BdataRWC,1)
tail(blc[,1:5],1)
BdataRWC1 <- BdataRWC[-nrow(BdataRWC),]
BdataBLC <- BdataRWC1 # creating an intermediate step so no rows are accidentally deleted

BdataBLC[,4:ncol(BdataBLC)] <- BdataBLC[,4:ncol(BdataBLC)] - blc[rep(1,104),4:ncol(blc)]

# xylose <- BdataRWC1[1,4] - blc[,4] # double check that the values are correct
# mb39 <- BdataRWC1[104,715] - blc[,715]

# Create the general dataset
GLDS138 <- BdataBLC

GLDS138$Label <- gsub("B. subtilis", "", GLDS138$Label)
GLDS138$Label <- gsub("S. aureus", "", GLDS138$Label)
GLDS138$Label <- gsub("S. Aureus", "", GLDS138$Label)

# autoscale function
autoscale <- function(a){
  colmean <- apply(a,2,mean) # column mean
  colsd <- apply(a, 2, sd) # col standard deviation
  cv <- sweep(a, 2, colmean, "-") # minus mean center
  cv <- sweep(cv, 2, colsd, "/") # divide by sd
  return(cv)
}

# Normalise
GLDS138.as <- GLDS138
GLDS138.as[,4:ncol(GLDS138.as)] <- autoscale(GLDS138.as[,4:ncol(GLDS138.as)])

write.csv(GLDS138.as,"GLDS138.as.csv", row.names = F)
```
# Comparison setups 
```{r}
GLDS138.as <- read.csv("GLDS138.as.csv", check.names = FALSE, header = TRUE)
utils::View(GLDS138.as)

# Just B. subtilis
Bsub.as <- GLDS138.as[c(1:52),]
Bsub.as <- Bsub.as[,-3]

# Get rid of Control ASAP and Frozen - so just media and pellet
BsubMP <- Bsub.as[-c(1:8, 27:34),]
BsubMP$Label <- factor(BsubMP$Label)

# Just Pellet - the bacteria - comparison of just flight vs ground
BsubFG <- BsubMP[c(19:36),]
BsubFG$Label <- gsub("Pellet ", "", BsubFG$Label)
# write.csv(BsubFG,"GLDS138.BsubFG.as.csv", row.names = F) # used later for metaboanalyst

# Just Media consumed by the bacteria - F vs G
BsubMd <- BsubMP[c(1:18),]
BsubMd$Label <- gsub("Media ", "", BsubMd$Label)

```
# Start analysis of PCA
```{r}
library(mixOmics)
# the general file
X <- GLDS138.as[,4:ncol(GLDS138.as)]
MyResult.pca <- pca(X, ncomp = 3, center = TRUE)
plotIndiv(MyResult.pca, comp = c(1,2), legend = TRUE, title = 'GLDS138: PCA',
          group = GLDS138.as$Label, legend.title = 'Treatment',
          ellipse = FALSE, style = "ggplot2", ind.names = FALSE, 
          pch = GLDS138.as$species, legend.title.pch = 'Species')
plotVar(MyResult.pca)
plotVar(MyResult.pca, cutoff = 0.8)
MyResult.pca #34,21,7

# B subtilis
X <- Bsub.as[,3:ncol(Bsub.as)]
MyResult.pca <- pca(X, ncomp = 3, center = TRUE)
plotIndiv(MyResult.pca, comp = c(1,2), legend = TRUE, title = 'B. subtilis: PCA',
          group = Bsub.as$Label, legend.title = 'Treatment',
          ellipse = TRUE, style = "ggplot2", ind.names = FALSE)
plotVar(MyResult.pca, cutoff = 0.8)
MyResult.pca #32,28,7

# Media and pellet
X <- BsubMP[,3:ncol(BsubMP)]
MyResult.pca <- pca(X, ncomp = 3, center = TRUE)
plotIndiv(MyResult.pca, comp = c(1,2), legend = TRUE, title = 'B. subtilis - Media vs Pellet: PCA',
          group = BsubMP$Label, legend.title = 'Treatment',
          ellipse = TRUE, style = "ggplot2", ind.names = FALSE)
plotVar(MyResult.pca, cutoff = 0.8)
MyResult.pca #33,11,8

# Pellet - Flight vs Ground
X <- BsubFG[,3:ncol(BsubFG)]
MyResult.pca <- pca(X, ncomp = 4, center = TRUE)
plotIndiv(MyResult.pca, comp = c(1,2), legend = TRUE, title = 'B. subtilis - Flight vs Ground: PCA',
          group = BsubFG$Label, legend.title = 'Treatment',
          ellipse = TRUE, style = "ggplot2", ind.names = FALSE)
plotIndiv(MyResult.pca, comp = c(1,2,3), legend = TRUE, title = 'B. subtilis - Flight vs Ground: PCA',
          group = BsubFG$Label, legend.title = 'Treatment',
          ellipse = TRUE, style = "3d", ind.names = FALSE)
plotVar(MyResult.pca, cutoff = 0.8)
MyResult.pca #18,15,11,11

# Media - Flight vs Ground
X <- BsubMd[,3:ncol(BsubMd)]
MyResult.pca <- pca(X, ncomp = 3, center = TRUE)
plotIndiv(MyResult.pca, comp = c(1,2), legend = TRUE, title = 'B. subtilis consumed media - Flight vs Ground: PCA',
          group = BsubMd$Label, legend.title = 'Treatment',
          ellipse = TRUE, style = "ggplot2", ind.names = FALSE)
plotVar(MyResult.pca, cutoff = 0.8)
MyResult.pca #28,13,12

# top metabolites of PC1
#metab_scores <- abs(MyResult.pca$rotation[,1])
#metab_score_ranked <- sort(metab_scores, decreasing = TRUE)
#top_10_metab <- names(metab_score_ranked[1:10])
#MyResult.pca$rotation[top_10_metab,1]

top_10_metab <- names(sort(abs(MyResult.pca$rotation[,1]), decreasing = TRUE)[1:10])
MyResult.pca$rotation[top_10_metab,1]

# top metabolites for PC2
top_10_metab <- names(sort(abs(MyResult.pca$rotation[,2]), decreasing = TRUE)[1:10])
MyResult.pca$rotation[top_10_metab,2]

# top metabolites for PC3
top_10_metab <- names(sort(abs(MyResult.pca$rotation[,3]), decreasing = TRUE)[1:10])
MyResult.pca$rotation[top_10_metab,3]
```
# sPCA
```{r}
# B subtilis
X <- Bsub.as[,3:ncol(Bsub.as)]
MyResult.spca <- spca(X, ncomp = 3, center = TRUE, keepX = c(16,14,5))
plotIndiv(MyResult.spca, comp = c(1,2), legend = TRUE, title = 'B. subtilis: sPCA',
          group = Bsub.as$Label, legend.title = 'Treatment',
          ellipse = TRUE, style = "ggplot2", ind.names = FALSE)
plotVar(MyResult.spca, cex = 3)

# Media and pellet
X <- BsubMP[,3:ncol(BsubMP)]
MyResult.spca <- spca(X, ncomp = 3, center = TRUE, keepX = c(17,10,5))
plotIndiv(MyResult.spca, comp = c(1,2), legend = TRUE, title = 'B. subtilis - Media vs Pellet: sPCA',
          group = BsubMP$Label, legend.title = 'Treatment',
          ellipse = TRUE, style = "ggplot2", ind.names = FALSE)
plotVar(MyResult.spca, cex = 3)

# Pellet - Flight vs Ground
X <- BsubFG[,3:ncol(BsubFG)]
MyResult.spca <- spca(X, ncomp = 3, center = TRUE, keepX = c(18,15,12))
plotIndiv(MyResult.spca, comp = c(1,2), legend = TRUE, title = 'B. subtilis - Flight vs Ground: sPCA',
          group = BsubFG$Label, legend.title = 'Treatment',
          ellipse = TRUE, style = "ggplot2", ind.names = FALSE)
plotVar(MyResult.spca, cex = 3)

# Media - Flight vs Ground
X <- BsubMd[,3:ncol(BsubMd)]
MyResult.spca <- spca(X, ncomp = 3, center = TRUE, keepX = c(15,10,6))
plotIndiv(MyResult.spca, comp = c(1,2), legend = TRUE, title = 'B. subtilis consumed media - Flight vs Ground: sPCA',
          group = BsubMd$Label, legend.title = 'Treatment',
          ellipse = TRUE, style = "ggplot2", ind.names = FALSE)
plotVar(MyResult.spca, cex = 3)
```
# Anova
```{r}
anova_data <- BsubFG[,c("Label", "210885", "110647", "120996", "101299", "5085", "stearic acid")]
library(tidyverse)
anova_data %>% pivot_longer(-Label, names_to = "Metabolite", values_to = "value") %>%
  mutate(Metabolite = factor(Metabolite, levels = c("210885", "110647", "120996", "101299", "5085", "stearic acid"))) %>%
  ggplot(aes(x = Label, y = value, color = Metabolite)) +
  geom_boxplot() +
  facet_wrap(~Metabolite)

aov.models = anova_data[ , -grep("Label", names(anova_data))] %>%
  map(~ aov(.~ anova_data$Label))
summary(aov.models$`210885`)
summary(aov.models$`110647`)
summary(aov.models$`120996`)
summary(aov.models$`101299`)
summary(aov.models$`5085`)
summary(aov.models$`stearic acid`)
```
#Partial Least Squares - Discriminant Analysis
```{r}
# set up
X <- BsubFG[,-c(1:2)]
Y <- BsubFG$Label
summary(Y)
dim(X)
length(Y)

# PLS-DA
MyResult.plsda <- plsda(X,Y)
plotIndiv(MyResult.plsda)
plotVar(MyResult.plsda, cutoff = 0.75)

# sPLS-DA
MyResult.splsda <- splsda(X, Y, keepX = c(60,60))
plotIndiv(MyResult.splsda, ind.names = FALSE, legend = TRUE,
          ellipse = TRUE, star = TRUE, title = 'Spaceflight B. subtilis: sPLS-DA',
          X.label = 'PLS-DA 1', Y.label = 'PLS-DA 2')
plotVar(MyResult.splsda, cutoff = 0.85)
selectVar(MyResult.splsda, comp = 1)$name

# Background prediction
background <- background.predict(MyResult.splsda, comp.predicted = 2,
                                 dist = "max.dist")
plotIndiv(MyResult.splsda, comp = 1:2, group = BsubFG$Label,
          ind.names = FALSE, title = "Maximum distance",
          legend = TRUE,  background = background)

# ROC - check for ability to predict classification
aud.plsda <- auroc(MyResult.splsda)

# Variable selection output
MyResult.splsda2 <- splsda(X, Y, ncomp = 3, keepX = c(38,19,9))
selectVar(MyResult.splsda2, comp = 1)$value
plotLoadings(MyResult.splsda2, contrib = 'max', method = 'mean')
plotIndiv(MyResult.splsda2, style="3d")

# Anova test on pls-da's top variables
plsda_data <- BsubFG[,c("Label", "ribonic acid", "methionine", "110014", "223522", "128027", "224109")]
library(tidyverse)
plsda_data %>% pivot_longer(-Label, names_to = "Metabolite", values_to = "value") %>%
  mutate(Metabolite = factor(Metabolite, levels = c("ribonic acid", "methionine", "110014", "223522", "128027", "224109"))) %>%
  ggplot(aes(x = Label, y = value, color = Metabolite)) +
  geom_boxplot() +
  facet_wrap(~Metabolite)

#
aov.plsda.models = plsda_data[ , -grep("Label", names(plsda_data))] %>%
  map(~ aov(.~ plsda_data$Label))
summary(aov.plsda.models$`ribonic acid`)
summary(aov.plsda.models$`methionine`)
summary(aov.plsda.models$`110014`)
summary(aov.plsda.models$`223522`)
summary(aov.plsda.models$`128027`)
summary(aov.plsda.models$`224109`)
```

# MetaboAnalyst0
```{r}
setwd("C:/Users/strki/Documents/Git_R/GLDS-138 metabolomics/GLDS-138-metabolomics/Compre Bsub PCA")
library(MetaboAnalystR)

# set up
mSet<-InitDataObjects("pktable", "stat", FALSE)
mSet<-Read.TextData(mSet, "GLDS138.BsubFG.as.csv", "rowu", "disc");
mSet<-SanityCheckData(mSet)
mSet<-ReplaceMin(mSet);
mSet<-PreparePrenormData(mSet)
mSet<-Normalization(mSet, "NULL", "NULL", "NULL", ratio=FALSE, ratioNum=20)
mSet<-PlotNormSummary(mSet, "norm_0_", "png", 72, width=NA)
mSet<-PlotSampleNormSummary(mSet, "snorm_0_", "png", 72, width=NA)

# Anova
mSet<-ANOVA.Anal(mSet, F, 1.0E-3, "fisher", FALSE)
mSet<-PlotANOVA(mSet, "aov_fisher_", "png", 72, width=NA)
mSet<-ANOVA.Anal(mSet, F, 1.0E-3, "tukey", FALSE)
mSet<-PlotANOVA(mSet, "aov_tukey_", "png", 72, width=NA)
# appears to be the same/similar metabolites as with more variables

# Correlation analysis
mSet<-PlotCorrHeatMap(mSet, "corr_1_", "png", 102, width=NA, "row", "pearson", "topo", "overview", F, F, F, 999)

# Pattern Hunter
mSet<-Match.Pattern(mSet, "pearson", "1-2")
mSet<-PlotCorr(mSet, "ptn_1_", "png", 72, width=NA) 
mSet<-Match.Pattern(mSet, "pearson", "2-1")
mSet<-PlotCorr(mSet, "ptn_2_", "png", 72, width=NA)

# PCA
mSet<-PCA.Anal(mSet)
mSet<-PlotPCAPairSummary(mSet, "pca_pair_0_", "png", 102, width=NA, 5)
mSet<-PlotPCAScree(mSet, "pca_scree_0_", "png", 72, width=NA, 5)
mSet<-PlotPCA2DScore(mSet, "pca_score2d_0_", "png", 72, width=NA, 1,2,0.95,0,0)
mSet<-PlotPCALoading(mSet, "pca_loading_0_", "png", 72, width=NA, 1,2);
mSet<-PlotPCABiplot(mSet, "pca_biplot_0_", "png", 102, width=NA, 1,2)
mSet<-PlotPCA3DLoading(mSet, "pca_loading3d_0_", "json", 1,2,3)

# Hierarchical Clustering
mSet<-PlotHCTree(mSet, "tree_0_", "png", 72, width=NA, "euclidean", "ward.D")
mSet<-PlotHeatMap(mSet, "heatmap_0_", "png", 300, width=NA, "norm", "row", "euclidean", "ward.D","bwm", "overview", T, T, NA, T, F)

# PLS-DA
mSet<-PLSR.Anal(mSet, reg=TRUE)
mSet<-PlotPLSPairSummary(mSet, "pls_pair_0_", "png", 72, width=NA, 5)
mSet<-PlotPLS2DScore(mSet, "pls_score2d_0_", "png", 72, width=NA, 1,2,0.95,0,0)
mSet<-PlotPLS3DScoreImg(mSet, "pls_score3d_0_", "png", 72, width=NA, 1,2,3, 40)
mSet<-PlotPLSLoading(mSet, "pls_loading_0_", "png", 72, width=NA, 1, 2);
mSet<-PlotPLS3DLoading(mSet, "pls_loading3d_0_", "json", 1,2,3)
mSet<-PLSDA.CV(mSet, "L",4, "Q2")
mSet<-PlotPLS.Classification(mSet, "pls_cv_0_", "png", 72, width=NA)
mSet<-PlotPLS.Imp(mSet, "pls_imp_0_", "png", 72, width=NA, "vip", "Comp. 1", 15,FALSE)

#Kmeans
mSet<-Kmeans.Anal(mSet, 4)
mSet<-PlotKmeans(mSet, "km_1_", "png", 150, width=NA)

# Random Forest
mSet<-RF.Anal(mSet, 1000,7,1)
mSet<-PlotRF.Classify(mSet, "rf_cls_1_", "png", 72, width=NA)
mSet<-PlotRF.VIP(mSet, "rf_imp_1_", "png", 72, width=NA)
mSet<-PlotRF.Outlier(mSet, "rf_outlier_1_", "png", 72, width=NA)

# Enrichment analysis
mSet<-InitDataObjects("conc", "msetora", FALSE)
cmpd.vec<-c("ribonic acid","isothreonic acid","isocitric acid","pyruvic acid","tryptophan","methionine")
mSet<-Setup.MapData(mSet, cmpd.vec);
mSet<-CrossReferencing(mSet, "name");
mSet<-CreateMappingResultTable(mSet)
mSet<-PerformDetailMatch(mSet, "isothreonic acid");
mSet<-GetCandidateList(mSet);
mSet<-SetCandidate(mSet, "isothreonic acid", "Threonic acid");
mSet<-SetMetabolomeFilter(mSet, F);
mSet<-SetCurrentMsetLib(mSet, "snp", 2);
mSet<-CalculateHyperScore(mSet)
mSet<-PlotORA(mSet, "ora_0_", "net", "png", 72, width=NA)
mSet<-CalculateHyperScore(mSet)
mSet<-PlotORA(mSet, "ora_1_", "net", "png", 72, width=NA)

# Pathway Analysis
mSet<-InitDataObjects("conc", "pathora", FALSE)
cmpd.vec<-c("ribonic acid","isothreonic acid","isocitric acid","pyruvic acid","tryptophan","methionine")
mSet<-Setup.MapData(mSet, cmpd.vec);
mSet<-CrossReferencing(mSet, "name");
mSet<-CreateMappingResultTable(mSet)
mSet<-PerformDetailMatch(mSet, "isothreonic acid");
mSet<-GetCandidateList(mSet);
mSet<-SetCandidate(mSet, "isothreonic acid", "Threonic acid");
mSet<-SetKEGG.PathLib(mSet, "bsu", "current")
mSet<-SetMetabolomeFilter(mSet, F);
mSet<-CalculateOraScore(mSet, "rbc", "hyperg")
mSet<-PlotPathSummary(mSet, "path_view_0_", "png", 72, width=NA)
```
# Set up for Multi-omics
```{r}
# Set up transcriptomic data
setwd("C:/Users/strki/Documents/Git_R/GLDS-138 metabolomics/GLDS-138-metabolomics/Compre Bsub PCA")
file_url <- "https://genelab-data.ndc.nasa.gov/genelab/static/media/dataset/GLDS-138_rna_seq_Normalized_Counts.csv?version=1"
GLDS138RNA = 'GLDS-138_rna_seq_Normalized_Counts.csv'
download.file(file_url, GLDS138RNA)
edata <- read.csv(GLDS138RNA, stringsAsFactors = FALSE)

dim(edata)
rowName <- edata[,1]
rownames(edata) <- rowName
edata <- edata[,-1]
colnames(edata) <- substr(colnames(edata),start=16,stop=23)
edata <- as.data.frame(edata)
edata <- t(edata)
rowNameE <- rownames(edata)

# set up metabolomic data
mdat <- BsubFG[,-c(1:2)]
rownames(mdat) <- rowNameE
# mdata$Label <- as.factor(mdata$Label)

# set up lipidomics data
Ldata <- read.csv("GLDS138.Lipid.as.csv", row.names = 1, check.names = FALSE)
Ldata <- Ldata[c(19:36),]
Ldata <- Ldata[,-c(1:2)]
rownames(Ldata) <- rowNameE
```
# Projection to Latent Structure (PLS)
```{r}
#quick start on PLS
X <- edata
Y <- mdat
dim(X); dim(Y)

MyResult.spls <- spls(X, Y, keepX = c(20, 20), keepY = c(10,10))
plotIndiv(MyResult.spls, group = BsubFG$Label)
plotVar(MyResult.spls, cex = c(3,3), legend = TRUE)

plotIndiv(MyResult.spls, group = BsubFG$Label,
          rep.space = "XY-variate", legend = TRUE,
          legend.title = 'Treatment',
          title = 'Spaceflight Bacillus subtilis: sPLS')

# Clustered image maps
X11()
cim(MyResult.spls, comp = 1)

#Relevance networks
X11()
network(MyResult.spls, comp = 1, interactive = TRUE)
# network(MyResult.spls, comp = 1, cutoff = 0.89, save = 'jpeg',name.save = 'PLS.network') # if intending to save it

# Arrow plots
plotArrow(MyResult.spls,group=BsubFG$Label, legend = TRUE,
          X.label = 'PLS comp 1', Y.label = 'PLS comp 2')
# Short arrows indicate a good agreement found by the PLS between both data sets
# more useful if we're looking at time related, or soemthing linking from one treatment to another.

# Variable selection outputs
MySelectedVariables <- selectVar(MyResult.spls, comp = 1)
MySelectedVariables$X$name # Selected genes on component 1

MySelectedVariables$Y$name # Selected metabolites on component 1

plotLoadings(MyResult.spls, comp = 1, size.name = rel(0.5))
```

# Multi-block Discriminant Analysis with DIABLO
```{r}
library(mixOmics)
X <- list(RNA = edata,
          metabolite = mdat,
          lipid = Ldata)
Y <- BsubFG$Label
summary(Y)

list.keepX <- list(RNA = c(10,6), metabolite = c(10,6), lipid = c(10,6))

# quick start
MyResult.diablo <- block.splsda(X, Y, keepX=list.keepX)
plotIndiv(MyResult.diablo,
          ind.names = TRUE,
          legend = TRUE, cex = c(1,2),
          title = "Spaceflight B. subtilis with DIABLO") ## sample plot
plotVar(MyResult.diablo, var.names = c(FALSE, TRUE, FALSE),
        legend = TRUE, pch = c(16, 1.5, 2))
plotDiablo(MyResult.diablo, ncomp = 1)

circosPlot(MyResult.diablo, cutoff = 0.93, ncol.legend = 1, showIntraLinks = TRUE, line = TRUE, size.variables = 0.6)
cimDiablo(MyResult.diablo, color.blocks = c('darkorchid', 'brown1', 'lightgreen'), comp = 1, margin=c(8,20), legend.position = "right")
plotLoadings(MyResult.diablo, comp = 1, contrib = "max", size.name = 1.25)
plotLoadings(MyResult.diablo, comp = 2, contrib = "max", size.name = 1.25)
par(mfrow=c(1,1))

# Relevance network
network(MyResult.diablo, blocks = c(1,2,3),
        color.node = c('darkorchid', 'brown1', 'lightgreen'), 
        cutoff = 0.6, save = 'jpeg', name.save = 'DIABLOnetwork')

MyResult.diablo2 <- block.plsda(X, Y) 
```



















```{r}
devtools::session_info()
```